### DISCUSSION TITLE
How can I find duplicate images by percent similarity in Linux?

### DISCUSSION THREAD

**Discussion by imthenachoman** [ImageMagick/ImageMagick#6440] [Help]
### ImageMagick version

7.1.1-11

### Operating system, version and so on

Linux

### Description

I have read https://www.imagemagick.org/Usage/compare but I can't figure this out.

I have a folder with thousands of pictures. I'm trying to find all duplicates based on percent similarity.

Ideally I'd like to dump the data to a text file so I can view each duplicate file path for manual analysis.

**Comment by fmw42** [comment]
Are the image all the same dimensions?  Are they all color or are some grayscale?

**Reply by imthenachoman** [reply]
No. They are all color. Some are black and white. It's a huge mix of photos we've taken with a DSLR and phones over the last decade.

**Reply by fmw42** [reply]
You say all are color. The you say some are b/w.  That is inconsistent.

Also you did not answer the question about whether they are all the same size.

Furthermore, how would you consider them different, but acceptable. What differences?

**Reply by imthenachoman** [reply]
> Are the image all the same dimensions?

No

> Are they all color or are some grayscale?

Mix of color and black/white.

> how would you consider them different, but acceptable. What differences?

I'd like to specify some sort of % similar threshold.

These are all photos we've taken over the years. Sometimes we made a copy of the photo that is resized smaller, or lower quality. I consider those as duplicates. 

That's what I want to find, those "duplicate" photos.

**Reply by fmw42** [reply]
This is mostly a scripting issue.  But the basic comparison might be done using perceptual hash, which is built into Imagemagick.  See http://www.fmwconcepts.com/misc_tests/perceptual_hash_test_results_510/index.html

**Reply by imthenachoman** [reply]
I'm having trouble understanding the output of `compare -metric phash image1 image2 diffimage`. It spits out a # but what does that # mean? How can I conver that # to a percent or something I can use to identify similarity based on percent threshold.

**Reply by fmw42** [reply]
phash metric computes the sum of the squared differences of all 42 or 7 log(abs(moments)) between the two images

Smaller is better.  There is no maximum.  So it does not convert easily to a percent.

If you want a percent use -metric NCC, but then the two images must be the same dimensions.

**Reply by imthenachoman** [reply]
Same exact dimensions or same ratio? I will look into NCC. My "similar" photos won't have the same dimensions. They would be resized versions mostly.

**Reply by fmw42** [reply]
No other metric but phash works on dissimilar dimension images unless looking for a subset. So they would not work for scale or rotation changes.
